### ISSUE TITLE
Build a fair evaluation of the prior

### ISSUE THREAD

**Issue by rom1504** [lucidrains/DALLE2-pytorch#29]
We're starting to have our first decent prior now. (PR of the training script coming soon)

Time to evaluate
Ideas:
* Mse on test set
* Zero shot eval on image net Class Text -> text emb -> image emb -> ranking https://github.com/LAION-AI/project-menu/issues/13
* Clip guided generation with the prior

If you have more ideas please share, I may be missing some obvious things.

We have more volunteers that want to help so I'll point some here :)

**Comment by lucidrains** [comment]
Already?! 🙏💯🎉

**Comment by lucidrains** [comment]
I was going to add training scripts for all the components tomorrow morning ... 😂

**Comment by lucidrains** [comment]
Eager to figure out just how important this prior network really is 🤔

**Comment by lucidrains** [comment]
I think for the clip guided generation, you will still need a decoder conditioned on the clip image embedding, tho you can probably get away with a small resolution net for starters, just to validate

**Comment by xiankgx** [comment]
Thank you for your work! I've modified your code to work with the original CLIP and got the training scripts. What is a good small dataset to test this on?

**Comment by lucidrains** [comment]
@xiankgx 👋 are you working with Laion? you should, because they have a humongous (and smaller test) datasets

**Comment by lucidrains** [comment]
> Thank you for your work! I've modified your code to work with the original CLIP and got the training scripts. What is a good small dataset to test this on?

Do you have a link to the CLIP that you used? I can try to incorporate it tomorrow using https://github.com/lucidrains/DALLE2-pytorch/blob/main/dalle2_pytorch/dalle2_pytorch.py#L95

**Comment by xiankgx** [comment]
> > Thank you for your work! I've modified your code to work with the original CLIP and got the training scripts. What is a good small dataset to test this on?
> 
> Do you have a link to the CLIP that you used? I can try to incorporate it tomorrow using https://github.com/lucidrains/DALLE2-pytorch/blob/main/dalle2_pytorch/dalle2_pytorch.py#L95

I am using the CLIP from this link: https://github.com/openai/CLIP

**Comment by lucidrains** [comment]
ohh got it, looks like they finally got it to be pip installable. I'll take a look tomorrow at an adapter!

**Comment by lucidrains** [comment]
ok, the plan will be to automatically use the openai clip by setting a `use_openai_clip` on both the `Decoder` and `DiffusionPrior`, which will allow researchers to skip the first step in the whole process

**Comment by xiankgx** [comment]
> @xiankgx wave are you working with Laion? you should, because they have a humongous (and smaller test) datasets

How can I help? Would be glad to help.

**Comment by lucidrains** [comment]
> > > Thank you for your work! I've modified your code to work with the original CLIP and got the training scripts. What is a good small dataset to test this on?
> > 
> > 
> > Do you have a link to the CLIP that you used? I can try to incorporate it tomorrow using https://github.com/lucidrains/DALLE2-pytorch/blob/main/dalle2_pytorch/dalle2_pytorch.py#L95
> 
> I am using the CLIP from this link: https://github.com/openai/CLIP

https://github.com/lucidrains/DALLE2-pytorch/tree/0.0.67#openai-clip ok, should be OpenAI clip compatible now, at some point I'll make it OpenCLIP compatible as well

**Comment by rom1504** [comment]
you can depend on clip-anytorch if you want openai clip from pypi (that's my pypi deployment of it)

**Comment by lucidrains** [comment]
nice! I'll refactor to use it maybe next week :)

**Comment by rom1504** [comment]
Oh I mean there is no change to be made except pip install clip-anytorch
Everything else is the same

**Comment by lucidrains** [comment]
@rom1504 it works! :pray: https://github.com/lucidrains/DALLE2-pytorch/releases/tag/0.0.73

**Comment by rom1504** [comment]
https://huggingface.co/rom1504/dalle2-diffusion-prior/resolve/main/1651432174.5708027_saved_model.pth here's a first checkpoint for the prior

let's start evaluation work!

**Comment by rom1504** [comment]
https://colab.research.google.com/drive/1kUYIvWje6CVO9llqY_9bYYk6zMNh1sSh?usp=sharing first eval from theo
comparing image predicted with real embedding
not amazing

let's include that kind of metrics in the training to see if things improve with time

**Comment by TheoCoombes** [comment]
I should note here, there could be a good chance this could be some form of normalization issue on my end causing the results to be worse as this is my first time playing around with the DALLE2-pytorch code

**Comment by lucidrains** [comment]
@TheoCoombes so one thing to note is that in the paper, they actually sampled a couple image embeddings (well, just 2 i guess), and then selected the one with the highest similarity to the text embedding. so it seems they must have encountered the same difficulties. the logic is in this function here if you need it! https://github.com/lucidrains/DALLE2-pytorch/blob/main/dalle2_pytorch/dalle2_pytorch.py#L831

**Comment by rom1504** [comment]
https://colab.research.google.com/drive/10P81dVS7YKCMUHF3FA7WD3Q_mp-cCWIA#scrollTo=VVElbFFcb5T7 new eval with the new checkpoint https://huggingface.co/krish240574/Dalle2-Diffusion-Prior/blob/main/1651473037.600823_saved_model.pth

now it works, we improve the similarity!
previous checkpoint was 0.27 -> 0.09
now 0.27 -> 0.28

we're going to make a PR to evaluate that automatically during training since it's cheap

**Comment by lucidrains** [comment]
@rom1504 very nice! :D

**Comment by nousr** [comment]
I have a new model/eval set. This run tries out optimization parameters that are more in line with what the paper specifies.

It seems to show better performance, resulting in a similarity of ~0.78 (up from 0.28, on the previously used image). However, more work should be done on benchmarking since early testing in the discord shows that unrelated prompts can also score relatively high similarities.

Here is the model repository [(hugging face link)](https://huggingface.co/zenglishuci/diffusion-prior-25M/tree/main).
And a W&B [report](https://wandb.ai/nousr_laion/prior-25M/reports/Similarity-Metrics--VmlldzoxOTMzNTYy?accessToken=04h92ryfmrbbuu9t4txqn6tm6t78gf80bubm0lzlx2s75xydux5dr7lbjtqtukk4) of a 25M datapoint run.

**Comment by nousr** [comment]
I've got a 300M point run going with the improved norm (re: #60).

I've also attempted to add a way to track the similarity with an unrelated text embedding. In short I shuffle the text embeddings in an effort to simulate "unrelated" prompts...

I'll PR the code when the run finishes and have you guys take a look at it to make sure the code & run results match what we would expect.

You can keep an eye on the run here [(wandb report)](https://wandb.ai/nousr_laion/diffusion-prior/reports/Improved-Norm-Prior-New-Metric--VmlldzoxOTU1NjI2?accessToken=rredjadr811vuq3j2jynlt89yn1gsqnvzi81md2m3moe2bd6ifhrt6ngrbgmhl1c)

**Comment by NasirKhalid24** [comment]
I trained a prior with ViT-B/32 text and image embeddings using the train_diffusion_prior.py script. Additionally i tracked the score for a fixed text embedding vs predicted image embedding. Ideally it should decrease over time since the text and images will be unrelated - it does but not as much and my weights toward the end still give a high score for unrelated text and predicted image embeds

CosineSim(Unrelated_Text_Embed, Prior(Text)) is upper graph

CosineSim(Related_Image_Embed, Prior(Text)) is lower graph
 


<img width="333" alt="Screen Shot 2022-05-06 at 1 56 10 PM" src="https://user-images.githubusercontent.com/29706203/167193498-8662040e-f639-474b-8cbb-bb4bc5011526.png">

**Comment by xiankgx** [comment]
Based on experience with CLIP, many texts can have the same cosine similarity ballpark even if some texts are better than others. Perhaps we can use softmax accuracy instead between generated image embeddings and input (text embeddings).

**Comment by nousr** [comment]
I just launched a run to test out a new metric, using deep-image-prior (#59), to generate images from the diffusion prior at 10k step intervals during training.

You can keep an eye on it here https://wandb.ai/nousr_laion/diffusion-prior/reports/Diffusion-Prior-With-D-I-P--VmlldzoxOTk2OTgy?accessToken=xmurn9eyp80bbliogmtsqatz2ot26eg1fp80sbsxukggllcb6n3z8dspbqo39cr1

If we get to ~50k steps and it looks like it works, then we can create a much more diverse prompt-set to evaluate the prior on.

**Comment by lucidrains** [comment]
https://github.com/lucidrains/DALLE2-pytorch/issues/23#issuecomment-1127011855

**Comment by rom1504** [comment]
this is almost done, zero shot eval might be the last thing here, some people are on it
